File: Python/device_information.py
import platform
import getpass
import psutil
import GPUtil
import time


class SystemInfo:
    def __init__(self):
        self.informacoes = {}
        self.collect_info()
    
    def collect_info(self):
        print("\nCarregando as informações...")
        # - Nome do sistema
        self.informacoes["Nome do Sistema"] = platform.system()
        # - Versão do sistema
        self.informacoes["Versão do Sistema"] = platform.platform()
        # - Nome do processador
        self.informacoes["Nome do Processador"] = platform.processor()
        # - Detalhes da arquitetura
        arquitetura = platform.architecture()
        self.informacoes["Arquitetura"] = f"{arquitetura[0]} | {arquitetura[1]}"
        # - Usuário do sistema
        self.informacoes["Usuário"] = getpass.getuser()
        # - Máquina
        self.informacoes["Máquina"] = platform.machine()
        # - Rede
        self.informacoes["Rede"] = platform.node()
        # - Uso da CPU (intervalo de 4 segundos)
        self.informacoes["Porcentagem da CPU"] = self.convert_value(psutil.cpu_percent(4)) + "%"
        # - Uso de memória RAM
        self.informacoes["Porcentagem de RAM"] = self.convert_value(psutil.virtual_memory().percent) + "%"
        # - Bateria
        bateria = psutil.sensors_battery()
        porcentagem_bateria = str(bateria.percent)
        carregando = bateria.power_plugged
        self.informacoes["Porcentagem de Bateria"] = self.convert_value(porcentagem_bateria) + "%"
        self.informacoes["Modo de Energia"] = "ligado à tomada" if carregando else "ligado à bateria"
        # - Discos
        self.informacoes["Discos"] = self.disk_info()
        # - Placas de Vídeo
        self.informacoes["Placas de Vídeo"] = self.gpu_info()
        # - Tempo de Atividade
        self.informacoes["Tempo de Atividade"] = f"{self.convert_value(self.get_uptime())} horas"
    
    # Informações sobre o disco
    def disk_info(self):
        disk_info = []
        for particao in psutil.disk_partitions():
            usage = psutil.disk_usage(particao.mountpoint)
            disk_info.append({
                "Disco": particao.device,
                "Tipo": particao.fstype,
                "Espaço Total (GB)": self.convert_value(usage.total / (1024 ** 3)),
                "Espaço Livre (GB)": self.convert_value(usage.free / (1024 ** 3)),
                "Espaço Usado (GB)": self.convert_value(usage.used / (1024 ** 3)),
                "Uso (%)": self.convert_value(usage.percent)
            })
        return disk_info

    # Informações sobre a(s) GPU(s)
    def gpu_info(self):
        gpus = GPUtil.getGPUs()
        gpu_info = []
        for gpu in gpus:
            gpu_info.append({
                "ID": gpu.id,
                "Nome": gpu.name,
                "Memória Total (MB)": self.convert_value(gpu.memoryTotal),
                "Memória Livre (MB)": self.convert_value(gpu.memoryFree),
                "Memória Usada (MB)": self.convert_value(gpu.memoryUsed),
                "Uso (%)": self.convert_value(gpu.memoryUtil * 100)
            })
        return gpu_info

    # Obter tempo de funcionamento do sistema
    def get_uptime(self):
        boot_time = psutil.boot_time()
        uptime = time.time() - boot_time
        return uptime / 3600
    
    def convert_value(self, valor):
        return str(round(float(valor), 2)).replace(".", ",")

    def print_info(self):
        print("\nInformações:")
        for item, valor in self.informacoes.items():
            print(f"- {item}: {valor}")
            time.sleep(0.3)
        print("\nProcesso finalizado.")


# Instanciando e usando a classe
if __name__ == "__main__":
    system_info = SystemInfo()
    system_info.print_info()

File: NodeJS/device_information.js
const si = require('systeminformation');
const diskinfo = require('diskinfo');
const gpuInfo = require('gpu-info');
const os = require("os");


class SistemaInformacoes {
    constructor() {
        this.informacoes = {};
    }

    // Método para obter informações do sistema
    async obterInformacoes() {
        console.log("\nCarregando as informações...");
        const cpus = os.cpus();

        // Coleta de informações do sistema
        this.informacoes["Arquitetura"] = os.arch();
        this.informacoes["Plataforma"] = os.platform();
        this.informacoes["Processador"] = cpus[0].model;

        // Memória
        const memoria_total = (os.totalmem() / 1024) / 1024;
        this.informacoes["Memória RAM Total (GB)"] = memoria_total.toFixed(2);
        const memoria_livre = (os.freemem() / 1024) / 1024;
        this.informacoes["Memória Livre (GB)"] = memoria_livre.toFixed(2);
        const porcentagem = (memoria_livre / memoria_total) * 100;
        this.informacoes["Porcentagem de Memória Livre"] = `${porcentagem.toFixed(2)}%`;

        // Tempo da máquina ligada
        const tempo = (os.uptime() / 60) / 60;
        this.informacoes["Uptime"] = tempo.toFixed(2);

        // Interfaces de redes
        this.informacoes["Interfaces de Redes"] = this.obterInformacoesRedes();
        // Caminho do usuário
        this.informacoes["Caminho do Usuário"] = os.homedir();
        // Versão do Sistema Operacional
        this.informacoes["Versão do Sistema Operacional"] = os.release();
        // Sistema Operacional
        this.informacoes["Sistema Operacional"] = os.type();
        // Nome do Host
        this.informacoes["Hostname"] = os.hostname();

        // Informações sobre CPUs
        const cpuInfoList = cpus.map((cpu, index) => {
            return `\n-- CPU ${index + 1}:\nModelo: ${cpu.model}\nNúcleos: ${cpu.cores}\nTempo de Idle: ${cpu.times.idle} ns\nTempo de Usuário: ${cpu.times.user} ns\nTempo de Sistema: ${cpu.times.sys} ns`;
        });
        this.informacoes["Quantidade de CPU"] = cpus.length;
        this.informacoes["CPUs"] = cpuInfoList;

        // Informações sobre a bateria
        this.informacoes["Bateria"] = await this.obterInformacoesBateria();
        // Informações sobre o disco
        this.informacoes["Disco"] = await this.obterInformacoesDisco();
        // Informações sobre a GPU
        this.informacoes["GPU"] = await this.obterInformacoesGPU();
    }

    // Método para obter interfaces de rede
    obterInformacoesRedes() {
        const interfaces = os.networkInterfaces();
        const informacoesRedes = Object.keys(interfaces).map(interfaceName => {
            const infos = interfaces[interfaceName];
            return infos.map(info => 
                `\n-- Interface ${interfaceName}\nEndereço: ${info.address}\nTipo: ${info.family}\nIPv6: ${info.internal ? 'Interno' : 'Externo'}\n${info.mac ? `MAC: ${info.mac}` : ''}`
            );
        });
        return informacoesRedes;
    }

    // Método para obter informações da bateria
    async obterInformacoesBateria() {
        try {
            const battery = await si.battery();
            return [
                `\nPercentual: ${battery.percent}%`,
                `\nTempo Restante: ${battery.timeRemaining} minutos`,
                `\nStatus de Carregamento: ${battery.isCharging ? 'Carregando' : 'Descarregando'}`,
                `\nNível de Energia: ${battery.powerPlugged ? 'Plugado' : 'Desconectado'}`
            ];
        } catch (error) {
            console.error('Erro ao obter informações da bateria: ', error);
            return [];
        }
    }

    // Método para obter informações do disco
    async obterInformacoesDisco() {
        return new Promise((resolve, reject) => {
            diskinfo.getDrives((err, drives) => {
                if (err) {
                    reject('Erro ao obter informações de disco: ' + err);
                    return;
                }
                const informacoesDisco = drives.map(drive => 
                    `\nTipo do Disco ${drive.mounted}\nDisco: ${drive.filesystem}\nTotal: ${(drive.blocks / 1024 / 1024 / 1024).toFixed(2)} GB\nUsado: ${(drive.used / 1024 / 1024 / 1024).toFixed(2)} GB\nLivre: ${(drive.available / 1024 / 1024 / 1024).toFixed(2)} GB\nPercentual Usado: ${(drive.used / drive.blocks * 100).toFixed(2)}%`
                );
                resolve(informacoesDisco);
            });
        });
    }

    // Método para obter informações da GPU
    async obterInformacoesGPU() {
        try {
            const gpus = await gpuInfo();
            if (!Array.isArray(gpus)) {
                throw new Error('Dados de GPU inesperados.');
            }

            const informacoesGPU = gpus.map((gpu, index) => {
                // Garantir que as propriedades existam antes de acessá-las
                const model = gpu.model || 'Desconhecido';
                const vendor = gpu.vendor || 'Desconhecido';
                const totalMemory = gpu.memory?.total ? (gpu.memory.total / 1024 / 1024).toFixed(2) + ' MB' : 'N/A';
                const usedMemory = gpu.memory?.used ? (gpu.memory.used / 1024 / 1024).toFixed(2) + ' MB' : 'N/A';
                const freeMemory = gpu.memory?.free ? (gpu.memory.free / 1024 / 1024).toFixed(2) + ' MB' : 'N/A';
                const temperature = gpu.temperature ? gpu.temperature + ' °C' : 'N/A';

                return `\n-- GPU ${index + 1}:\nModelo: ${model}\nFabricante: ${vendor}\nMemória Total: ${totalMemory}\nMemória Usada: ${usedMemory}\nMemória Livre: ${freeMemory}\nTemperatura: ${temperature}`;
            });

            return informacoesGPU;
        } catch (err) {
            console.error('Erro ao obter informações da GPU:', err);
            return [];
        }
    }

    // Método para exibir as informações
    exibirInformacoes() {
        console.log("\nInformações:");
        for (const item in this.informacoes) {
            console.log(`- ${item}: ${this.informacoes[item]}`);
        }
    }
}


(async () => {
    const sistema = new SistemaInformacoes();
    await sistema.obterInformacoes();
    sistema.exibirInformacoes();
})();